Save and restore values from type environment. Don't copy it all

vaivaswatha opened the pull request:
Fixes #643.

Note: I don't think the change to TypeCache.ml is correct, I've just gotten it to type check. Since we aren't using type caching currently, and enabling and testing it too much work, I've ignored it. It must be fixed when type caching is fixed and enabled.

This should make the type checker linear in the size of the program as against quadratic today.

jjcnn commented on src/base/TypeUtil.ml:
```diff
@@ -65,23 +65,31 @@ module type MakeTEnvFunctor = functor (Q : QualifiedTypes) (R : Rep) -> sig
 
     (* Add to type environment *)
-    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> t
+    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> restore list
```
Should this not be `val addT : t -> R.rep TUIdentifier.t -> TUType.t -> restore`? 

From an abstract point of view a `restore` is an entirely opaque value, so it seems reasonable that one `restore` should correspond to one operation.

jjcnn commented on src/base/TypeUtil.ml:
```diff
@@ -65,23 +65,31 @@ module type MakeTEnvFunctor = functor (Q : QualifiedTypes) (R : Rep) -> sig
 
     (* Add to many type bindings *)
-    val addTs : t -> (R.rep TUIdentifier.t * TUType.t) list -> t
+    val addTs : t -> (R.rep TUIdentifier.t * TUType.t) list -> restore list
```
You could also have the return type be `restore` rather than `restore list` here. The `restore` object returned by the function corresponds to the reversal of all the additions, but as a single `restore` operation.

jjcnn replied:
Similar observations about all the other `TEnv` functions.

jjcnn commented on src/base/TypeUtil.ml:
```diff
@@ -147,33 +152,50 @@ functor
+        | RemT of string
+        (* | AddV of string * R.rep *)
+        | RemV of string
+
```
If you go with the `restore` rather than `restore list` suggestion, then you will need something like

```
type restore_op =
  | AddT of ...
  etc.

type restore = restore_op list

type additional_restore r rs = r :: rs
```

The result will be that the from the point of view of the user of `TEnv`, there will just be one `restore` object for whichever operation they have performed, and there is no way to inspect or manipulate them (at the moment the restoration can be manipulated by removing `restore` objects from the returned lists).

vaivaswatha commented:
@jjcnn 

1. I have incorporated your suggestion about making `restore` entirely opaque.
2. I have removed TypeCache.ml(i) files, and a commented piece of code in TypeChecker.ml related to type caching.

Please see the latest two commits.

Thank you :)

vaivaswatha commented:
Also please ignore the TravisCI failure due to formatting. I'll fix it when I merge #859 to this branch before merging it all to `master`.

anton-trunov commented:
I liked the line count for this PR a lot :)

vaivaswatha commented:
> I liked the line count for this PR a lot :)

ha ha, that's because I deleted TypeCache.ml(i)
